streams/shared/src/main/scala/zio/stream/ZTransducer.scala

```diff
@@ -731,6 +778,64 @@ object ZTransducer {
       }
     }
 
+  /**
+   * Decodes chunks of UTF-16 bytes into strings.
+   *
+   * This will fail with an `IllegalArgumentException` if no byte order mark was found and will
+   * use the error handling behavior of the endian-specific decoder otherwise.
+   */
+  val utf16Decode: ZTransducer[Any, IllegalArgumentException, Byte, String] =
+    magicSequence(2) { bytes =>
+      bytes.toList match {
+        case -2 :: -1 :: Nil =>
+          utf16BEDecode
+        case -1 :: -2 :: Nil =>
+          utf16LEDecode
+        case xs =>
+          fail(new IllegalArgumentException(s"Not a valid magic sequence ${xs.mkString(", ")}"))
+      }
+    }
+
+  /**
+   * Decodes chunks of UTF-16 bytes into strings.
+   *
+   * This transducer uses the String constructor's behavior when handling malformed byte
+   * sequences.
+   */
+  val utf16BEDecode: ZTransducer[Any, Nothing, Byte, String] =
+    utf16Decode(StandardCharsets.UTF_16BE)
+
+  /**
+   * Decodes chunks of UTF-16 bytes into strings.
+   *
+   * This transducer uses the String constructor's behavior when handling malformed byte
+   * sequences.
+   */
+  val utf16LEDecode: ZTransducer[Any, Nothing, Byte, String] =
+    utf16Decode(StandardCharsets.UTF_16LE)
+
+  private def utf16Decode(charset: Charset): ZTransducer[Any, Nothing, Byte, String] =
+    ZTransducer {
+      ZRef.makeManaged[Option[Byte]](None).map { stateRef =>
+        {
+          case None =>
+            // should we raise an error here if we have leftovers?
```

Comment:
What should we do here? If we do not fail the transducer we will lose data as there is no way to decode the orphaned byte

Reply:
With the UTF-8 decoding, if the stream ends (which is what `None` signifies) in the middle of 2/3/4 byte sequence, we just defer to the underlying charset decoder (which if I recall correctly just drops the malformed bytes). I'm fine with adopting the same approach here.

In the future, the `None` case will really signal the end of stream and we will stop allowing transducers to be restarted during stream processing. This will be done by separating transducers for which restarts do make sense (all the folds basically) to a separate abstraction (`ZAggregator` or something like that).

Reply:
Ah, I see. Thanks for the explanation 👍